File: DifferentWaysToAddParentheses/DifferentWaysToAddParentheses.py
class Solution(object):
    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """
        if input.isdigit():
        	return [int(input)]
        result = []
        for i in xrange(len(input)):
        	cur = input[i]
        	if not cur.isdigit():
        		result1 = self.diffWaysToCompute(input[:i])
        		result2 = self.diffWaysToCompute(input[i+1:])
        		for i in result1:
        			for j in result2:
        				if cur == '+':
        					result.append(i + j)
        				elif cur == '-':
        					result.append(i - j)
        				else:
        					result.append(i * j)
        return result

File: DifferentWaysToAddParentheses/DifferentWaysToAddParentheses.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string input) {
        vector<int> result;
        for(int i = 0; i < input.size(); i++) {
        	char cur = input[i];
        	if(ispunct(cur)) {
        		vector<int> result1 = diffWaysToCompute(input.substr(0, i));
        		vector<int> result2 = diffWaysToCompute(input.substr(i + 1));
        		for(int a : result1)
        			for(int b : result2)
        				result.push_back(cur=='+' ? a + b:cur=='-'? a-b : a*b);
        	}
        }
        if(result.empty())
        	result.push_back(atoi(input.c_str()));
        return result;
    }
};